src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py

```diff
@@ -65,39 +69,106 @@ def generate(
 
         component_graph = connection_manager.get().component_graph
 
-        battery_inverters = frozenset(
-            frozenset(
+        # all_connected_batteries has to be set because
+        # multiple inverters can be connected to the same battery
+        inv_bat_mapping: dict[Component, set[Component]] = {}
+        all_connected_batteries: set[Component] = set()
+        for bat_id in component_ids:
+            inverters = set(
                 filter(
                     component_graph.is_battery_inverter,
                     component_graph.predecessors(bat_id),
                 )
             )
-            for bat_id in component_ids
-        )
-
-        if not all(battery_inverters):
-            raise ComponentNotFound(
-                "All batteries must have at least one inverter as a predecessor."
-            )
+            if len(inverters) == 0:
+                raise ComponentNotFound(
+                    "All batteries must have at least one inverter as a predecessor.",
+                    "Battery ID %s has no inverter as a predecessor.",
+                    bat_id,
+                )
```

Comment:
It is interesting that we need this type of checks in the code.  I'd have thought this is already done by the component graph.

Anyway, this is not a log statement, so the `%s` won't be replaced, you need to use an f-string.

Reply:
> Anyway, this is not a log statement, so the %s won't be replaced, you need to use an f-string.

Done

> It is interesting that we need this type of checks in the code. I'd have thought this is already done by the component graph.

I didn't check if component graph has it.